### PR TITLE
Use climate forecasts, provide all weather data, closes #216, closes …

### LINKED ISSUE #216: Change how we handle weather data lags
We should probably change the type of weather data that's provided in data. In future, models will want other lags besides 6 months (and won't be able to call the get weather data function if they're not written in R), and certainly the option for no lag. We should either:

1. provide the original data and let the individual model scripts deal with rearranging for lags

2. provide a file with the original data along with a couple choices of lags as additional columns

3. just provide the original data, improve the get weather data function a bit and let individual models use it to get lags, don't worry about non-R model scripts for now

Option 1 seems like it would be fine to me. . .

### PR DESCRIPTION
**Pull request by gmyenni**
…#249

* Provide all weather data in data folder
* Create a lag function that models can use to lag weather data relative to rodent data
* Create a weather forecast function that complies a 12 month forecast based on actual station data, longterm climate forecasts, and 5 year means to fill in missing values

### CODE REVIEW

**Comment by gmyenni** [comment]
@DAPPERstats , thought you should probably check this out and make sure pevGARCH is still working the way you want it to.

**Comment by juniperlsimonis** [comment]
If you're ok with it, @gmyenni, there are some edits I'd like to make to the lag_weather_data() function:
1. replace the camel case of "NewMoonNumber_with_lag" to match the all lowercase of "newmoonnumber" to be "newmoonnumber_with_lag"
2. have the output dataframe now have "year" and "month" columns that match "newmoonnumber" and then "year_with_lag" and "month_with_lag" columns that match "newmoon_number_with_lag" so that it's clear which year and month belongs to the lag vs the original data 

edit: actually, at that point, it kind of seems like it would make more sense to return a data set that has the "newmoonnumber" column that is lagged, but don't include the "_with_lag" tag, and not return the un-lagged times. that way, the functions that use the data can be flexible to whether the weather data are lagged or not (without requiring a flag to indicate the "_with_lag" tag should be added)

[also, I guess I'm in a funny linguistic mood right now with whether the weather and flagging lag tags]

**Comment by juniperlsimonis** [comment]
also, to address the issue of missing recent censuses, i propose we save out the moondata, which gets updated with the forecast date, and then use it within pevGARCH.
i've got code to execute that pretty easily in another branch and can bring it over here.

**Comment by juniperlsimonis** [comment]
that actually leads to another set of edits in lag_weather_data, to use the moon data (rather than the rodent data) in that function
i can make those edits, just letting you know.

edit: the updated portalr functions producing weather and ndvi data directly with newmoon numbers and the to-be updated pevGARCH function means that we actually can do the lag simply with the data set to be lagged and the amount to lag it by (i.e. no need for some additional data set like rodent data or moon data)

**Comment by juniperlsimonis** [comment]
I just pushed a bunch of work in a single commit here.
It will not pass functionality tests because it relies on upstream changes to portlar (https://github.com/weecology/portalr/pull/108), which need to be merged in there first.
Once the upstream changes have been merged in and the version of portalr is updated in the builds (is that done automatically?), the code should work. (it currently works locally with the updated version of portalr I built on my computer)

**Comment by ethanwhite** [comment]
>  the version of portalr is updated in the builds (is that done automatically?)

We need to do this manually. We do this intentionally to avoid breaking changes messing up the forecasting repo before we've had a chance to incorporate them. Once  https://github.com/weecology/portalr/pull/108 is in ping @gmyenni & me and one of us will test locally, rebuild the container, push it, and then restart this build for testing.

File: tools/model_tools.R [hunk 1]
```diff
@@ -39,3 +42,281 @@ interpolate_abundance <- function(abundance){
...
+
+  if(max_hist_ndvi_nm < last_newmoon){
+
+    newmoons_to_fcast <- (max_hist_ndvi_nm + 1):last_newmoon
+    ndvi_fcast <- fcast_ndvi(hist_ndvi, moons, newmoons_to_fcast, lag, FALSE)
+    hist_ndvi <- rbind(hist_ndvi, ndvi_fcast)  
+  }
+  return(hist_ndvi)
+}
+
```
Comment (gmyenni) [inline tools/model_tools.R]: Yeah yeah! Let's use this as our `fill` argument for ndvi over in portalr.

Comment (juniperlsimonis) [reply tools/model_tools.R]: just fyi, i'm going to need to generalize the function more, as right now it's based on newmoons only, and the ndvi can be managed as newmoons or months. it shouldn't be too much work, but it's more than just plopping the function over here.

**Comment by juniperlsimonis** [comment]
additional to do with the updates in https://github.com/weecology/portalr/pull/108
remove ndvi fill and forecast functions (now in portalr)
remove get_new_moons function (now in portalr)
ensure the names of and arguments to the ndvi functions in pevGARCH are still appropriate with the updates to the functions

**Comment by juniperlsimonis** [comment]
edit: wait, hold on, there are still some edits to be made in the prepare data script.

**Comment by juniperlsimonis** [comment]
ok @gmyenni , now the updates have been made in the prepare data script to use the new portlar functions and the four models have all been set up to forecast past dates when necessary (like right now, when the most recent newmoons did not have completed surveys)...but to use known (rather than forecasted...although in the case of the ndvi right now, it's actually still forecasted) covariates for those data. this required an alteration of the model meta data yaml file to now distinguish between forecast dates for the rodents and forecast dates for the covariates.

File: tools/prepare_data.R [hunk 2]
```diff
@@ -8,25 +10,52 @@ forecast_date = Sys.Date()
...
+first_forecast_covariate_newmoon <- most_recent_covariate_newmoon + 1
+first_forecast_rodent_newmoon <- most_recent_rodent_newmoon + 1
+covariate_forecast_newmoons <- first_forecast_covariate_newmoon:last_forecast_newmoon
+rodent_forecast_newmoons <- first_forecast_rodent_newmoon:last_forecast_newmoon
+covariate_nm_dates <- total_moons$newmoondate[total_moons$newmoonnumber %in% covariate_forecast_newmoons]
+rodent_nm_dates <- total_moons$newmoondate[total_moons$newmoonnumber %in% rodent_forecast_newmoons]
+covariate_forecast_months <- as.numeric(format(covariate_nm_dates, "%m"))
+covariate_forecast_years <- as.numeric(format(covariate_nm_dates, "%Y"))
+rodent_forecast_months <- as.numeric(format(rodent_nm_dates, "%m"))
+rodent_forecast_years <- as.numeric(format(rodent_nm_dates, "%Y"))
```
Comment (gmyenni) [inline tools/prepare_data.R:44]: I know what the problem is here, and the need for rodent dates and covariate dates is still confusing. I think we might need more explanation.

**Pull request merged**